### ISSUE TITLE
Enhanced nav: support redirections triggered by non-Blazor code

### ISSUE THREAD

**Issue by SteveSandersonMS** [dotnet/aspnetcore#53606]
Currently, during an enhanced nav POST request, redirections to external URLs will only work if they are triggered by `NavigationManager.NavigateTo(url)`. They do **not** work if triggered through any other API, such as `httpContext.Response.Redirect`. This is a problem because other ASP.NET middleware may issue redirections, for example to auth endpoints.

### Customer impact

Blazor enhanced forms are often used to perform data access, such as saving changes via a POST, or simpler gestures like "add to cart" in eShop. But it's always possible that your authentication token has expired and that some non-Blazor middleware is going to intercept the request and perform a redirection to an external URL for reauthentication.

Right now there's no sensible way to manage this other than simply not enabling enhancement on the form. Unless you can 100% guarantee that no other middleware might issue a redirection, you just can't use `Enhance` or `data-enhance`, because any such redirection will lead to an error.

We could make redirections work regardless of what code triggered them.

### Proposed solution

Blazor should, in some way, hook into the ASP.NET Core request processing pipeline (via middleware or any other mechanism) so that it can detect the case of "enhanced nav request leads to a redirection" and perform the same response translation that Blazor's renderer does for `NavigationException`.

The logic would look the same as we have in EndpointHtmlRenderer.Prerendering.cs: https://github.com/dotnet/aspnetcore/blob/5d0e3b0513a1fd71aa4a63029b91d57504b59345/src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.Prerendering.cs#L203-L213

The perf impact would be negligible since it starts by checking for method=post and the presence of the enhanced nav marker header, and if either is absent it's a no-op.

### Note

We considered doing this in .NET 8 but decided not to take the risk of affecting non-Blazor endpoints so close to the final release. For .NET 9 we have more opportunity to attempt it.

**Comment by javiercn** [comment]
The challenging bit here is to do this in a "transparent" way I think. I hope we don't end up with a middleware in the pipeline to deal with this, but the only alternatives I can think of is hooking up into [OnResponseStarting](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.httpresponse.onstarting?view=aspnetcore-8.0#microsoft-aspnetcore-http-httpresponse-onstarting(system-func((system-threading-tasks-task)))) to change the response accordingly.

Otherwise, things like the auth `Challenge` or code that sets the status code on the response won't work.

Maybe we can figure out a way in which we can wire up middleware conditionally to the default builder, I think that would be good enough.

At that point we would check for POST+enhanced nav header and attach a delegate to the context to apply the enhanced nav protocol if we see a redirect triggered.

I think that will be good enough as it won't be in the user's code by default and can be wired up explicitly if needed be.

**Comment by javiercn** [comment]
For how to wire up middleare implicitly check [src/DefaultBuilder/src/WebApplicationBuilder.cs](https://github.com/dotnet/aspnetcore/blob/main/src/DefaultBuilder/src/WebApplicationBuilder.cs#L389)

**Comment by SteveSandersonMS** [comment]
> We do fall back to a regular navigation on the client when this happens, don't we?

I don't think so - like you mention, the client doesn't get to find out what the redirection target URL is. And since it's a form post, it can't retry the post as it's not idempotent.

**Comment by garrettlondon1** [comment]
Currently I am using a combination of Blazor Web pages and full SSR (no blazor.web) pages together for full JS support without PageScript (using blazor component model).

The way I get around this for the JS pages:

```
HttpContextAccessor.HttpContext.Response.Headers["Blazor-Enhanced-Nav"] = "false";
```

To tell Blazor that when redirecting, don't use enhanced nav for the DOM swap, but this will force the client to make two calls as noted above..

It seems the solution to this with Blazor on the client is to add the disable-enhanced-nav on the A or NavLink? But it leaves developers to manually handle each tag, with the error being hard to catch and significant

Is there a different thread where this is being discussed? Is the goal of this thread to not get it to make two requests? Or specifically regarding form post idempotency
